### PR TITLE
ViTMatte (Part 1 of MatteAnything)

### PR DESCRIPTION
**Pull request by chloedia**
ViT Matte PR for the implementation of Matte Anything :

- [x] ViT backbone
- [x] Decoder
- [x] Conversion script

### CODE REVIEW

File: scripts/conversion/convert_ViTMatte.py
Comment (limiteinductive) [inline scripts/conversion/convert_ViTMatte.py:1]: nit: convert_ViTMatte.py -> convert_vit_matte.py

**Review by limiteinductive — COMMENTED** [review]
Running the VitMatte conversion script I have the following error:

```
RuntimeError:
   File "/home/benjamin/refiners/src/refiners/foundationals/ViT_matte/vit_backbone.py", line 127, in forward
    horizontal_relative_embedding, vertical_relative_embedding = self.compute_relative_embedding(x=query)
  File "/home/benjamin/refiners/src/refiners/foundationals/ViT_matte/vit_backbone.py", line 154, in compute_relative_embedding
    horizontal_relative_embedding = torch.einsum("bhwc,wkc->bhwk", x, horizontal_positional_embedding).unsqueeze(
  File "/home/benjamin/refiners/.venv/lib/python3.10/site-packages/torch/functional.py", line 380, in einsum
    return _VF.einsum(equation, operands)  # type: ignore[attr-defined]
einsum(): subscript w has size 14 for operand 1 which does not broadcast with previously seen size 64
---------------
(CHAIN) FusedSelfAttention(embedding_dim=384, num_heads=6)
    ├── Linear(in_features=384, out_features=1152, device=cpu, dtype=float32) #1
    ├── >>> RelativePositionAttention(embedding_dim=384, num_heads=6, spatial_size=(14, 14), device=cpu, dtype=float32) | Transformer.TransformerLayer_1.Residual_1.FusedSelfAttention.RelativePositionAttention
    └── Linear(in_features=384, out_features=384, device=cpu, dtype=float32) #2
0: Tensor(shape=(9, 14, 14, 1152), dtype=float32, device=cpu, min=-2.71, max=2.66, mean=0.00, std=0.44, norm=627.90, grad=False)
---------------
(RES) Residual()
    ├── LayerNorm(normalized_shape=(384,), eps=1e-06, device=cpu, dtype=float32)
    ├── WindowPartition()
    ├── >>> (CHAIN) FusedSelfAttention(embedding_dim=384, num_heads=6) | Transformer.TransformerLayer_1.Residual_1.FusedSelfAttention
    │   ├── Linear(in_features=384, out_features=1152, device=cpu, dtype=float32) #1
    │   ├── RelativePositionAttention(embedding_dim=384, num_heads=6, spatial_size=(14, 14), device=cpu, dtype=float32)
    │   └── Linear(in_features=384, out_features=384, device=cpu, dtype=float32) #2
    └── WindowMerge()
0: Tensor(shape=(9, 14, 14, 384), dtype=float32, device=cpu, min=-4.76, max=4.31, mean=-0.00, std=0.76, norm=627.07, grad=False)
---------------
(CHAIN) TransformerLayer(embedding_dim=384, num_heads=6, feedforward_dim=1536, window_size=14, image_embedding_size=(32, 32))
    ├── >>> (RES) Residual() | Transformer.TransformerLayer_1.Residual_1 #1
    │   ├── LayerNorm(normalized_shape=(384,), eps=1e-06, device=cpu, dtype=float32)
    │   ├── WindowPartition()
    │   ├── (CHAIN) FusedSelfAttention(embedding_dim=384, num_heads=6)
    │   │   ├── Linear(in_features=384, out_features=1152, device=cpu, dtype=float32) #1
    │   │   ├── RelativePositionAttention(embedding_dim=384, num_heads=6, spatial_size=(14, 14), device=cpu, dtype=float32)
    │   │   └── Linear(in_features=384, out_features=384, device=cpu, dtype=float32) #2
    │   └── WindowMerge()
    └── (RES) Residual() #2
        ├── LayerNorm(normalized_shape=(384,), eps=1e-06, device=cpu, dtype=float32)
0: Tensor(shape=(1, 32, 32, 384), dtype=float32, device=cpu, min=-5.40, max=4.85, mean=-0.00, std=1.05, norm=655.82, grad=False)
---------------
(CHAIN) Transformer()
    ├── >>> (CHAIN) TransformerLayer(embedding_dim=384, num_heads=6, feedforward_dim=1536, window_size=14, image_embedding_size=(32, 32)) (x2) | Transformer.TransformerLayer_1 #1
    │   ├── (RES) Residual() #1
    │   │   ├── LayerNorm(normalized_shape=(384,), eps=1e-06, device=cpu, dtype=float32)
    │   │   ├── WindowPartition()
    │   │   ├── (CHAIN) FusedSelfAttention(embedding_dim=384, num_heads=6) ...
    │   │   └── WindowMerge()
    │   └── (RES) Residual() #2
    │       ├── LayerNorm(normalized_shape=(384,), eps=1e-06, device=cpu, dtype=float32)
    │       └── (CHAIN) FeedForward(embedding_dim=384, feedforward_dim=1536) ...
    ├── (CHAIN) TransformerLayer(embedding_dim=384, num_heads=6, feedforward_dim=1536, image_embedding_size=(32, 32)) #2
0: Tensor(shape=(1, 32, 32, 384), dtype=float32, device=cpu, min=-5.40, max=4.85, mean=-0.00, std=1.05, norm=655.82, grad=False)
---------------
(CHAIN) MViTH()
    ├── (CHAIN) PatchEncoder(in_channels=4, out_channels=384)
    │   ├── Conv2d(in_channels=4, out_channels=384, kernel_size=(16, 16), stride=(16, 16), device=cpu, dtype=float32)
    │   └── Permute(dims=(0, 2, 3, 1))
    ├── (RES) PositionalEncoder(embedding_dim=384, num_position=197, image_embedding_size=(32, 32))
    │   ├── Parameter(dims=(197, 384), device=cpu, dtype=float32)
    │   └── Lambda(get_abs_pos(x: torch.Tensor) -> torch.Tensor)
    ├── >>> (CHAIN) Transformer() | .Transformer
    │   ├── (CHAIN) TransformerLayer(embedding_dim=384, num_heads=6, feedforward_dim=1536, window_size=14, image_embedding_size=(32, 32)) (x2) #1
    │   │   ├── (RES) Residual() #1 ...
    │   │   └── (RES) Residual() #2 ...
0: Tensor(shape=(1, 32, 32, 384), dtype=float32, device=cpu, min=-5.40, max=4.85, mean=-0.00, std=1.05, norm=655.82, grad=False)
```

**Comment by chloedia** [comment]
@limiteinductive error should be fixed !

File: src/refiners/foundationals/ViT_matte/decoder.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+                fl.Chain(
+                    VitMatteBasicConv3x3(self.conv_chans[i], self.conv_chans[i + 1], device=device, dtype=dtype),
+                    fl.SetContext("detailed_features", f"{i+1}"),
+                )
+                for i in range(len(self.conv_chans) - 1)
+            ],
+        )
+
+
+class Fusion_Block(fl.Chain):
```
Comment (limiteinductive) [inline src/refiners/foundationals/ViT_matte/decoder.py:63]: nit: Python convention is PascalCase for class names, aka. FusionBlock

File: src/refiners/foundationals/ViT_matte/decoder.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+            fl.Concatenate(
+                fl.Chain(fl.UseContext(context="detailed_features", key=self.conv_key)),
+                fl.Interpolate(factor=2, mode="bilinear"),
+                dim=1,
+            ),
+            VitMatteBasicConv3x3(self.in_channels, self.out_channels, stride=1, padding=1),
+        )
+
+
+class Matting_Head(fl.Chain):
```
Comment (limiteinductive) [inline src/refiners/foundationals/ViT_matte/decoder.py:86]: ditto

File: src/refiners/foundationals/ViT_matte/decoder.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+                dtype=dtype,
+            ),
+        )
+
+
+class FusionBlocks(fl.Chain):
+    pass
+
+
+class Detail_Capture(fl.Chain):
```
Comment (limiteinductive) [inline src/refiners/foundationals/ViT_matte/decoder.py:117]: ditto

File: src/refiners/foundationals/ViT_matte/vit_backbone.py
Comment (limiteinductive) [inline src/refiners/foundationals/ViT_matte/vit_backbone.py:1]: ViT_matte -> vit_matte

File: src/refiners/foundationals/ViT_matte/vit_backbone.py [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+                fl.Permute(0, 2, 3, 1),
+            )
+        )
+
+
+class Transformer(fl.Chain):
+    pass
+
+
+# -----------------------------------------------------------------------
```
Comment (limiteinductive) [inline src/refiners/foundationals/ViT_matte/vit_backbone.py:367]: to be removed

File: src/refiners/foundationals/ViT_matte/vit_backbone.py [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+
+
+class MViT(fl.Chain):
```
Comment (limiteinductive) [inline src/refiners/foundationals/ViT_matte/vit_backbone.py:370]: -> Let's name it ViTMatte

File: src/refiners/foundationals/ViT_matte/vit_backbone.py [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+                    device=device,
+                    dtype=dtype,
+                )
+                for i in range(num_layers)
+            ),
+            fl.Permute(0, 3, 1, 2),
+        )
+
+
+class MViTH(MViT):
```
Comment (limiteinductive) [inline src/refiners/foundationals/ViT_matte/vit_backbone.py:420]: nit: ViTMatteH

File: src/refiners/foundationals/ViT_matte/vit_matte.py [hunk 7]
```diff
@@ -0,0 +1,11 @@
+import refiners.fluxion.layers as fl
+from refiners.foundationals.ViT_matte.decoder import Detail_Capture
+from refiners.foundationals.ViT_matte.vit_backbone import MViTH
+
+
+class ViTMatte(fl.Chain):
```
Comment (limiteinductive) [inline src/refiners/foundationals/ViT_matte/vit_matte.py:6]: Could this be merged with the "MViT" class? Or both of them are needed? If it's the case, the name should reflect their specific function.

File: src/refiners/foundationals/ViT_matte/vit_matte.py
Comment (limiteinductive) [inline src/refiners/foundationals/ViT_matte/vit_matte.py:1]: nit: I don't like tautological path aka. using twice vit_matte ; you can vite_matte.py -> model.py if this module is really needed.

**Review by limiteinductive — COMMENTED** [review]
Could you write some unit tests?

**Pull request closed without merging**